### DISCUSSION TITLE
addTools to pasted Link

### DISCUSSION THREAD

**Discussion by Ndmanh123** [clientIO/joint#1699] [Q&A]
I have a use self.clipboard.pasteCells to paste 2 cell and a link but i cant seem to add tools to the link that paste is alway log me this err 
![image](https://user-images.githubusercontent.com/104048582/173527079-7f3892ff-106b-4b04-9f38-32e853dfb665.png)
have any one know how to fix this problem 
Thank

**Comment by kumilingus** [comment]
Could you provide steps to reproduce?

**Reply by Ndmanh123** [reply]
i use ctrl x to get the cell

```js
keyboard.on('keydown:ctrl+x', (evt) => {
                self.clipboard.cutElements(
                    self.selection.collection,
                    paper.model,
                );
            });
```            
and

```js
keyboard.on('keydown:ctrl+v', (evt) => {
                var pastedCells = self.clipboard.pasteCells(graph, {
                    translate: { dx: 40, dy: 40 },
                    useLocalStorage: true,
                });
                var elements = _.filter(pastedCells, function (cell) {
                    return cell.isElement();
                });
                // Make sure pasted elements get selected immediately. This makes the UX better as
                // the user can immediately manipulate the pasted elements.
                self.selection.collection.reset(elements);
})
```
here is ctrl V i listen to graph.on add to add all the tools is work fine with others link bit the new link that create by ctrl v is run in to bug

**Reply by kumilingus** [reply]
What is the code that listens to the `graph` and  adds the link tools?
The code you have sent are not steps to reproduce. That code works just fine.

**Reply by Ndmanh123** [reply]
```js
graph.on('add', (cell, collection, opt) => {
    let type = cell.attributes.type;
    if (type == 'app.Link') {
        let linkView = cell.findView(paper);
        var removeButton = new joint.linkTools.Remove({
            distance: '30%',
        });
        var verticesTool = new joint.linkTools.Vertices();
        var segmentsTool = new joint.linkTools.Segments();
        var targetArrowheadTool = new joint.linkTools.TargetArrowhead();
        var toolsView = new joint.dia.ToolsView({
            tools: [
                verticesTool,
                segmentsTool,
                removeButton,
                targetArrowheadTool,
            ],
        });

        linkView.addTools(toolsView);
        linkView.hideTools();
    }
});
```
 here is the code for add link tools

**Reply by Ndmanh123** [reply]
Thank i have fix that bug i just put setTimeOut in graph add when i add tool to link.
I think because the link when graph add is call is not render full in dom so is get bug when i add tool 
Hope this could help some one get the same problem

**Reply by kumilingus** [reply]
The `linkTools` tool could be added to cells that are not yet fully rendered (they will be rendered after the cellView is updated).

Using `setTimeout()` is rarely a solution. I wouldn't recommend it unless you know why you're doing it.

I tested your code using JointJS+ v3.5 and it works as expected. Unfortunately, the above code are not the steps to reproduce the problem.
